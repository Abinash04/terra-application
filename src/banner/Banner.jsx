import React from 'react';
import PropTypes from 'prop-types';
import uuidv4 from 'uuid/v4';

import BannerRegistrationContext from './_BannerRegistrationContext';
import { BANNER_TYPES } from './_Utils';

const propTypes = {
  /**
   * An action element to be added to the action section of the banner.
   */
  action: PropTypes.element,
  /**
   * Nodes providing the message content for the banner. Can contain text and HTML.
   */
  description: PropTypes.node,
  /**
   * Callback function triggered when Dismiss button is clicked. The presence of this prop will cause
   * the Dismiss button to be included on the banner.
   */
  onDismiss: PropTypes.func,
  /**
   * The title for the banner which will be bolded. Providing a title will override the default title.
   */
  title: PropTypes.string,
  /**
   * The type of alert to be rendered. One of `alert`, `error`, `warning`, `info`, `success` or `custom`.
   */
  type: PropTypes.oneOf([
    BANNER_TYPES.ALERT,
    BANNER_TYPES.ERROR,
    BANNER_TYPES.WARNING,
    BANNER_TYPES.INFO,
    BANNER_TYPES.SUCCESS,
    BANNER_TYPES.CUSTOM,
  ]).isRequired,
};

const Banner = (props) => {
  /**
   * A unique identifier is generated for each Banner during construction. This will be used to
   * uniquely register/unregister the banner with ancestor Banner Managers without requiring consumers to
   * define unique identifiers themselves.
   */
  const uuid = React.useRef(uuidv4());
  const bannerRegistration = React.useContext(BannerRegistrationContext);

  React.useEffect(() => {
    /**
     * If the bannerRegistration value is the ProviderRegistrationContext's default value,
     * then there is not a matching BannerCheckpoint above it in the hierarchy.
     * This is possible but likely not intentional, so the component warns.
     */
    if (!bannerRegistration && process.env.NODE_ENV !== 'production') {
      /* eslint-disable no-console */
      console.warn('A Banner was not rendered within the context of a BannerCheckpoint. If this is unexpected, validate that the expected version of the terra-application package is installed.');
      /* eslint-enable no-console */
    }

    bannerRegistration.registerBanner(uuid.current, props);

    return () => {
      bannerRegistration.unregisterBanner(uuid.current, props.type);
    };
  }, [props]);

  return null;
};

Banner.propTypes = propTypes;

export default Banner;
