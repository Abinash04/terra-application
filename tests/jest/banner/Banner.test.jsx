import React from 'react';
import { render } from '@testing-library/react';

import BannerRegistrationContext from '../../../src/banner/private/BannerRegistrationContext';
import Banner from '../../../src/banner';

const reactUseContext = React.useContext;

describe('Banner', () => {
  const registerBannerMock = jest.fn();
  const unregisterBannerMock = jest.fn();

  beforeAll(() => {
    React.useContext = (contextValue) => {
      if (BannerRegistrationContext === contextValue) {
        return {
          registerBanner: registerBannerMock,
          unregisterBanner: unregisterBannerMock,
        };
      }
      return reactUseContext(contextValue);
    };
  });

  afterAll(() => {
    // Restore Reacts useContext method after each test.
    React.useContext = reactUseContext;
  });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('registers an banner to the banner-list onmount', () => {
    render(<Banner type="alert" />);

    expect(registerBannerMock).toHaveBeenCalledTimes(1);
    expect(registerBannerMock).toHaveBeenCalledWith(expect.any(String), expect.objectContaining({ type: 'alert' }));
  });

  it('un-registers an banner from the banner-list onmount', () => {
    render(<Banner type="alert" />).unmount();

    expect(unregisterBannerMock).toHaveBeenCalledTimes(1);
    expect(unregisterBannerMock).toHaveBeenCalledWith(expect.any(String), 'alert');
  });
});
