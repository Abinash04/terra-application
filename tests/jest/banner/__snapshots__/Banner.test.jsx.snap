// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner registers an banner to the banner-list onmount 1`] = `
<Alert
  customColorClass="custom-default-color"
  type="alert"
/>
`;
